ntpd/src/daemon/system.rs

```diff
@@ -242,15 +257,31 @@ impl<C: NtpClock, T: Wait> System<C, T> {
         )
     }
 
-    fn add_spawner(&mut self, spawner: impl Spawner + Send + Sync + 'static) -> SpawnerId {
+    fn clock_controller(&mut self) -> Result<&mut KalmanClockController<C, PeerId>, C::Error> {
+        if self.controller.is_none() {
+            self.controller = Some(KalmanClockController::new(
+                self.clock.clone(),
+                self.synchronization_config,
+                self.peer_defaults_config,
+                self.synchronization_config.algorithm,
+            )?);
+        }
+        // Won't panic as the above if ensures controller contains something
+        Ok(self.controller.as_mut().unwrap())
+    }
```

Comment:
Interesting option that I didn't think about. In this case it won't work though, since the clock controller creation has non-trivial side effects that we really only want once, and get_or_insert takes a value, not a closure providing a value.

Perhaps the real solution here should be to somehow avoid needing side effects creating the clock controller, but it is not quite clear to me how that could work as it needs to do some initialization involving changes to the clock.

Reply:
a version that does not unwrap

```suggestion
        let controller = match self.controller.take() {
            Some(controller) => controller,
            None => KalmanClockController::new(
                self.clock.clone(),
                self.synchronization_config,
                self.peer_defaults_config,
                self.synchronization_config.algorithm,
            )?,
        };

        Ok(self.controller.insert(controller))
```

Reply:
Making a separate pull request for this as the suggestion needs a bit of testing.